README
#===============================================================================
# PyTyle - A manual tiling manager
# Copyright (C) 2009  Andrew Gallant <andrew@pytyle.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#===============================================================================

PyTyle is an extremely versatile and extensible tiling manager that is meant
to be embedded into EWMH window managers. Its feature set was modeled after
the basic tiling features of XMonad- even down to the key bindings.

It has been tested relatively well on OpenBox and partially on Fluxbox. It
should, in theory, work with any window manager complying with the EWMH
standards.

For related information, see:

http://pytyle.com

http://python-xlib.sourceforge.net/

http://tronche.com/gui/x/xlib/

http://tronche.com/gui/x/icccm/

http://users.actcom.co.il/~choo/lupg/tutorials/xlib-programming/xlib-programming.html

http://standards.freedesktop.org/wm-spec/wm-spec-1.3.html

CHANGELOG
#===============================================================================
# PyTyle - A manual tiling manager
# Copyright (C) 2009  Andrew Gallant <andrew@pytyle.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#===============================================================================

Last updated: Aug 28, 2009 at 0:17

See http://pytyle.com for more information. Also,
please don't hesitate to contact me:
    
    andrew@pytyle.com


#===============================================================================
0.7.4

- Bug fix: Detection of panel orientation
    If you let PyTyle automatically detect your panels (i.e., not setting your
    workarea), then it wasn't properly reserving space if your panel was at
    the top or left of your screen.
    
- Bug fix: Decorations (NEW CONFIGURATION OPTION ADDED)
    If you aren't using any decorations at all in Openbox, then PyTyle was still
    fiddling with them. Specifically, if you had decorations disabled on the
    Openbox end, then PyTyle would re-add them if you issued an untile command.
    
    Since Openbox still reports the sizes of decorations even when they're
    disabled, there really isn't a way to reliably detect whether a window has
    decorations or not. The solution here was to add a new configuration option:
    original_decor. This should simply be whether or not you have decorations
    enabled/disabled from the Openbox end.

#===============================================================================
0.7.3

- Detects new panels/docks on the fly
    PyTyle will now detect panels/docks added while it's running. (You shouldn't
    have to restart PyTyle when you change the orientation of your docks/panels.)
    
    Note: This will *only* work if you're using one screen and *haven't* set
    the WORKAREA option in your configuration file.
#===============================================================================
0.7.2

- New wiki added
    http://pytyle.com
    Check it out!

- Dynamically load your configuration file
    In keeping with the fashion of XMonad, you can now change your PyTyle
    configuration on the fly with the "reload" command (by default, it is
    assigned to the "Alt-Q" key binding). Anything goes. You could even add
    your own custom tiling layout from scratch while PyTyle is running.
    
    Note: The "global_tiling" option isn't available to be changed dynamically.
    Should it? Thoughts are welcome.
    
- Layout proportions are kept
    Before this version, whenever you added/removed a new window to your tiling
    layout, the master/slave proportions were reset to their defaults. This
    should not happen any more. Layouts now have the ability to save their
    "state" on the fly. (If you've been hacking your own layouts, it's important
    that you take a look at the code changes in, say, PyTyle/Tilers/Vertical.py
    so that you can incorporate saving state in your layout as well. It is very
    easy.)
    
    Note: You can reset your tiling state by doing one of the following three
    things:
        1. Change your tiling layout
        2. *Reset* your layout (This isn't the regular "Alt-A" but the
           "Alt-Shift-Space" command.)
        3. Reload your configuration file.
        
- More layout options
    Check out the default configuration file provided with this release for
    additional LAYOUT configuration options. Namely, you can change the default
    width/height factor of the Vertical/Horizontal layouts.
    
    Note: You do not need to incorporate these changes into your configuration
    file. PyTyle keeps all default config settings embedded to fall back on if
    a setting is missing in the configuration file.
    
    Note 2: If you ever want to reset your config file, you can do so easily by
    stopping PyTyle, renaming your current configuration file to something else,
    and starting PyTyle again. You will have a brand new default configuration. 
#===============================================================================
0.7.1

- Some bug fixes.    
#===============================================================================    
0.7

- Initial support for Compiz.
    This is done by detecting and calculating all of the viewports that
    Compiz uses instead of desktops. The only thing that's giving me a
    bit of trouble in Compiz is the Cascade layout, namely, something is
    going wrong when I'm resetting the stacking order. (Window focus is
    weird.) Also, if you're curious, x/y positions of windows in Compiz
    are *relative* to the current viewport. So if you're using wmctrl to
    debug, don't expect to get cogent x,y positions unless you figure in
    the current viewport (which wmctrl will give you with wmctrl -d).
    
    Note: Using multiple desktops with multiple viewports should, in theory
    work. It has not been tested.
    
    Note 2: PyTyle is only enabling viewport support when it detects that
    *Compiz* is the window manager. If any other window manager needs
    viewport support from PyTyle, I'll need to know about it to add it. (Or
    you could modify Probe.is_compiz.)
    
    Note 3: Per workspace/screen default layouts work with viewports. However,
    this will not work as expected if you're using viewports and multiple
    desktops.
    

PyTyle/Config.py
#===============================================================================
# PyTyle - A manual tiling manager
# Copyright (C) 2009  Andrew Gallant <andrew@pytyle.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#===============================================================================

"""
Config.py

A simple container class for the configuration of PyTyle. See pytylerc
for more information on configuring PyTyle.

The only variable here that isn't in pytylerc is Config.TILERS. This
configuration variable is loaded at program start up as a dict of all
available tiling algorithms (as Python modules).

There is also a getter method here, just in case we're looking for an
option that isn't specified in the config.
"""

class Config:
    #------------------------------------------------------------------------------
    # CONFIGURATION VARIABLES
    #------------------------------------------------------------------------------

    MISC = {}
    KEYMAP = {}
    WORKAREA = {}
    FILTER = []
    LAYOUT = {}
    TILING = {}
    TILERS = {}
    CALLBACKS = {}


    #------------------------------------------------------------------------------
    # CONFIGURATION DEFAULTS
    #------------------------------------------------------------------------------

    DEFAULTS = {
                'MISC': {
                         'tilers': ['Vertical', 'Horizontal', 'Maximal', 'Cascade'],
                         'global_tiling': False,
                         'timeout': 0.1,
                         'decorations': True,
                         'original_decor': True,
                         },
                'KEYMAP': {
                           'Alt-A': 'tile.default',
                           'Alt-U': 'untile',
                           'Alt-Z': 'cycle_tiler',
                           'Alt-Shift-space': 'reset',
                           'Alt-C': 'cycle',
                           'Alt-W': 'screen0_focus',
                           'Alt-E': 'screen1_focus',
                           'Alt-R': 'screen2_focus',
                           'Alt-Shift-W': 'screen0_put',
                           'Alt-Shift-E': 'screen1_put',
                           'Alt-Shift-R': 'screen2_put',
                           'Alt-H': 'master_decrease',
                           'Alt-L': 'master_increase',
                           'Alt-period': 'add_master',
                           'Alt-comma': 'remove_master',
                           'Alt-Return': 'make_active_master',
                           'Alt-M': 'win_master',
                           'Alt-Shift-C': 'win_close',
                           'Alt-J': 'win_previous',
                           'Alt-K': 'win_next',
                           'Alt-Shift-J': 'switch_previous',
                           'Alt-Shift-K': 'switch_next',
                           'Alt-X': 'max_all',
                           'Alt-S': 'restore_all',
                           },
                'WORKAREA': {
                             0: {
                                 'top': 0,
                                 'bottom': 0,
                                 'right': 0,
                                 'left': 0,
                                 },
                             },
                'FILTER': [
                           'gmrun', 'gimp', 'download',
                           ],
                'LAYOUT': {
                           'Vertical': {
                                        'width_factor': 0.5,
                                        'margin': 0,
                                        },
                           'Horizontal': {
                                          'height_factor': 0.5,
                                          'margin': 0,
                                          },
                           'Maximal': {},
                           'Cascade': {
                                       'decoration_height': 25,
                                       'width_factor': 1.0,
                                       'height_factor': 1.0,
                                       'push_over': 0,
                                       'horz_align': 'left',
                                       },
                           'HorizontalRows': {
                                              'row_size': 2,
                                              'height_factor': 0.5,
                                              'margin': 0,
                                              },
                           },
                'TILING': {
                           'default': 'Vertical',
                           },
                'CALLBACKS': {
                            0:'make_active_master',
                            1:'switch_previous',
                            2:'switch_next'
                            }
                }


    #------------------------------------------------------------------------------
    # CONFIGURATION GETTERS
    #------------------------------------------------------------------------------

    @staticmethod
    def misc(name):
        if name in Config.MISC:
            return Config.MISC[name]

        if name in Config.DEFAULTS['MISC']:
            return Config.DEFAULTS['MISC'][name]

        return None

    @staticmethod
    def keymap(keys):
        if keys in Config.KEYMAP:
            return Config.KEYMAP[keys]

        if keys in Config.DEFAULTS['KEYMAP']:
            return Config.DEFAULTS['KEYMAP'][keys]

        return None

    @staticmethod
    def workarea(screen_number, section):
        if screen_number in Config.WORKAREA and section in Config.WORKAREA[screen_number]:
            return Config.WORKAREA[screen_number][section]

        return Config.DEFAULTS['WORKAREA'][0][section]

    @staticmethod
    def filter():
        return Config.FILTER

    @staticmethod
    def layout(tiler, option):
        layout = tiler.__class__.__name__
        if layout in Config.LAYOUT and option in Config.LAYOUT[layout]:
            return Config.LAYOUT[layout][option]

        if layout in Config.DEFAULTS['LAYOUT'] and option in Config.DEFAULTS['LAYOUT'][layout]:
            return Config.DEFAULTS['LAYOUT'][layout][option]

        return None

    @staticmethod
    def tiling(screen, desk_or_view):
        if screen in Config.TILING:
            if isinstance(Config.TILING[screen], dict) and desk_or_view in Config.TILING[screen]:
                return Config.TILING[screen][desk_or_view]
            elif not isinstance(Config.TILING[screen], dict):
                return Config.TILING[screen]
        elif 'default' in Config.TILING:
            return Config.TILING['default']

        return Config.DEFAULTS['TILING']['default']

    @staticmethod
    def tilers(layout):
        if layout in Config.TILERS:
            return Config.TILERS[layout]

    @staticmethod
    def callbacks(num):
        if num in Config.CALLBACKS:
            return Config.CALLBACKS[num]
        if num in Config.DEFAULTS['CALLBACKS']:
            return Config.DEFAULTS['CALLBACKS'][num]
        return None

    # Special flag to enable/disable debugging
    #
    # PRIVACY NOTE: This may log the titles of
    # your windows!
    DEBUG = False
